src/hotspot/share/code/debugInfo.hpp

```diff
@@ -165,13 +187,59 @@ class ObjectValue: public ScopeValue {
   void print_fields_on(outputStream* st) const;
 };
 
+// An ObjectMergeValue describes objects that were inputs to a Phi in C2 and at
+// least one of them was scalar replaced.
+// '_selector' is an integer value that will be '-1' if during the execution of
+// the C2 compiled code the path taken was that of the Phi input that was NOT
+// scalar replaced. In that case '_merge_pointer' is a pointer to an already
+// allocated object. If '_selector' is not '-1' then it should be the index
+// of a candidate object in '_possible_objects'. That candidate object is an
+// ObjectValue describing an object that was scalar replaced.
+
+class ObjectMergeValue: public ObjectValue {
```

Comment:
I find the decision to subclass`ObjectValue` confusing and error prone: now `is_object()` returns true for `ObjectMergeValue`, but you have to apply the selector first to turn it into `ObjectValue`. And now the order of checks matter, so you always have to perform `is_object_merge()` first and then follow it with `is_object()` guard.

You have 3 flavors of `ObjectValue` now:
* good old `ObjectValue`;
* `ObjectMergeValue`
* merge candidates (`ObjectMergeCandidateValue`?)

Does it make sense to introduce 3 different subclasses under `ObjectValue` to clearly distinguish the scenarios?

Reply:
> Does it make sense to introduce 3 different subclasses under ObjectValue to clearly distinguish the scenarios?

I think that's a good idea. I'll give it a shot. Thanks.

Reply:
Can `ObjectCandidateValue` be a wrapper around a `ObjectAllocationValue`?

It does make sense to separate `ObjectMergeValue` and `ObjectValue`.

Reply:
I need to to study the code in more details. Seems like I'm missing something important here.

Reply:
@iwanowww - how can I make it easier for you to review? Thanks for your comments so far.